Fix the SSH issues related to the default SSH config and too many identity files

spott opened the pull request:
This PR addresses the following issues:

- #927
- #933
- #937

Implementation notes:

1. For issue 927, use `IdentitiesOnly` in the client SSH config
2. For issue 933, use `-F <client dstack SSH config path>` on the client, and `-F none` in other places when invoking the `ssh` command to exclude the default SSH config settings
3. For issue 937, display a clear error message

spott committed:
added -o IdentitiesOnly=yes to ssh command to prevent too many identities from causing problems with ssh

spott committed:
Added /build to .gitignore

spott committed:
added IdentitiesOnly=yes to ssh config generation to allow for vscode integration when there are too many authentication attempts due to too many keys in the ssh agent

spott committed:
Added a catch and a message if the users ssh config is not writable.

spott committed:
Added `-F {config.dstack_ssh_config_path}` to every ssh call I could find in the python tree.

spott committed:
Added `-F none` to the ssh calls in go (assuming that this is for the agent running where ~/.dstack/ssh/config doesn't exist).

Egor-S commented on gateway/src/dstack/gateway/services/tunnel.py:
You should not change this file. It is not a part of a CLI (`src/dstack/_internal/cli`). `dstack-gateway` is a separate package deployed on a gateway instance. There is no custom SSH config

spott replied:
Should we change it to `-F none` if there is no custom SSH config?

Egor-S replied:
We can, but actually there isn't any config — it's a dedicated instance fully managed by dstack

spott replied:
I changed it to `-F none`.  Let me know if you think something else should be done.

Egor-S commented on runner/internal/gateway/ssh.go:
This code is not used anymore. `dstack-gateway` opens a connection to the `dstack-runner`

spott replied:
I'm not sure what I should do here... if it is unused, should I delete it?

Egor-S commented on src/dstack/_internal/server/services/ssh/__init__.py:
It's a server code and config manager is supposed to be used by CLI. I guess it's better to use `-F none` here

spott replied:
Done.

spott committed:
removed configmanager from server, as the custom ssh config is a CLI thing (not a server thing.

spott committed:
removed custom config from gatway/src/dstack/gateway/services/tunnel.py.  replaced it with "none" to isolate it from whatever system it is running on.

peterschmidt85 committed:
- [Bug]: The SSH tunnel fails because of a messy ~/.ssh/config #933
- [Bug]: The SSH tunnel fails with Too many authentication failures #927
- Pull request review

peterschmidt85 commented:
@spott I've pushed some changes. Please let me know if all looks good to you.
cc @Egor-S

peterschmidt85 commented on src/dstack/_internal/utils/ssh.py:
```diff
@@ -56,8 +56,15 @@ def include_ssh_config(path: PathLike, ssh_config_path: PathLike = default_ssh_c
         if include not in content:
-            with open(ssh_config_path, "w") as f:
-                f.write(include + content)
+            try:
```
@spott 
I created the corresponding issue: #937

peterschmidt85 reviewed (changes requested):
Left the comment about handling the permission error. Please let me know what you think.

Egor-S commented on src/dstack/_internal/utils/ssh.py:
```diff
@@ -56,8 +56,15 @@ def include_ssh_config(path: PathLike, ssh_config_path: PathLike = default_ssh_c
+                with open(ssh_config_path, "w") as f:
+                    f.write(include + content)
+            except PermissionError:
+                print(f"Couldn't write to {ssh_config_path} due to a permissions problem")
```
@peterschmidt85 
We've discussed that a logger is a better option there. But the default logging level is CRITICAL, so we need to change it in `src/dstack/_internal/cli/utils/common.py : configure_logging`

peterschmidt85 committed:
- [Bug] Handle permission error when adding `Include` to `~/.ssh/config` #937
- Using the default logger to display the permission error
- Updated the permission error message for more detail
- Updated the default logging level for the CLI from `CRITICAL` to `WARNING`

peterschmidt85 committed:
- [Bug] Handle permission error when adding `Include` to `~/.ssh/config` #937
- Moved `extra: {"markup:True"}` to the `logger.warning` level
